src/m3ninx/index/segment/fst/segment.go

```diff
@@ -612,6 +617,37 @@ func (r *fsSegment) docsNotClosedMaybeFinalizedWithRLock(
 	return index.NewIDDocIterator(retriever, pl.Iterator()), nil
 }
 
+func (r *fsSegment) encodedDocNotClosedMaybeFinalizedWithRLock(id postings.ID) (doc.EncodedDocument, error) {
+	// NB(r): Not closed, but could be finalized (i.e. closed segment reader)
+	// calling match field after this segment is finalized.
+	if r.finalized {
+		return doc.EncodedDocument{}, errReaderFinalized
+	}
+
+	// TODO(nate): support for thirdPartyReader? seems to only to be used for compaction which
+	// is unnecessary for encoded docs use case of read queries.
+
+	offset, err := r.docsIndexReader.Read(id)
+	if err != nil {
+		return doc.EncodedDocument{}, err
+	}
+
+	return r.docsEncodedDataReader.Read(offset)
```

Comment:
Hm, won't be able to have an encoded docs data reader on the FST segment itself. There is concurrent access to the FST segment. 

This will have to be passed in from the FST segment reader (and allocated when a FST segment reader is created).

Reply:
Actually, this is the `EncodedDataReader`, not the `EncodedDocumentReader`. The `EncodedDataReader` doesn't have a `doc.Field` backing slice. It's just performing a readonly operation on the byte slice it's initialized with from the offset provided on `Read`

Reply:
Put another way, `EncodedDataReader` turns a byte slice into an EncodedDocument whereas the `EncodedDocumentReader` turns an EncodedDocument into a regular Document (which is where the `doc.Field` backing slice is used)

Reply:
Ah true, makes sense.